Change: move sensitive information to secrets.cfg and private information to private.cfg

TrueBrain opened the pull request:
## Motivation / Problem

While working on the STUN patch (#9017), I wanted to store a secret on the client. The only place we currently have to do this, is `openttd.cfg`. This has a huge draw-back for several reasons:

- We ask for `openttd.cfg` in bug-reports, basically telling the world about the secret.
- People tend to share their `openttd.cfg` pretty easily with others, sharing their secret too.

In my case, this would mean that getting your hands on an `openttd.cfg` meant you could claim the invite-code, and do a hostile take-over of their server.

A closer look made us realise that storing secrets in `openttd.cfg` is already an issue: `default_company_pass`, `rcon_password`, etc. In general it is not wise to store this information in the same file as all other configuration, especially as it is shared this easily.

## Description

In result, I set out to find a way to create a `secrets.cfg` which contains secrets people should not be sharing. The name should make that obvious enough.

Additionally, as suggested by @rubidium42 , I also created a `private.cfg` to store things like:
- server-name
- client-name
- last-joined
- server-bind-addresses
- servers
- bans

Basically, everything with identifiable information. This information people shouldn't be sharing freely either, but aren't exactly secrets.

When you start the game the first time with/after this PR, while already having a settings file from before, it will automatically migrate the settings to the new files and remove it from `openttd.cfg`. This is a one time migration.

## Limitations

- These secrets cannot be sync'd over the network, or stored in the savegame. I think that goes without saying that they also shouldn't be sync'd over the network or stored in the savegame .. that would defeat the purpose of calling it a secrets/private ;) But the code also really doesn't allow it.
- I am unsure if adding a `IniFileVersion` is really needed. I could also just add a flag or indicator. Not sure if this would ever be used again in the future. But I need at least something to know you have been past this PR with your configuration .. so alternatives for this are more than welcome.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

rubidium42 commented on src/settings.cpp:
```diff
@@ -1744,6 +1751,12 @@ void SaveToConfig()
+
+	IniFile *secrets_ini = new IniFile();
+	secrets_ini->LoadFromDisk(_secrets_file, NO_DIRECTORY);
+	IniSaveSettings(secrets_ini, _secrets_settings, "patches", &_settings_newgame, false);
```
Why `"patches"`? That's the group that is removed from the settings a dozen lines before this. I know it gets practically not used as all the setting names are `a.b`, so there would be an `[a]` in the ini. Maybe something clearer like "secrets".

TrueBrain replied:
it is the exact same line as used for `_settings` a bit higher. I wanted to be sure they behave identical ;)

rubidium42 commented on src/settings.cpp:
```diff
@@ -2067,6 +2080,10 @@ const SettingDesc *GetSettingFromName(const char *name)
 		if (!SlIsObjectCurrentlyValid(sd->save.version_from, sd->save.version_to)) continue;
 		if (strcmp(sd->name, name) == 0) return sd.get();
 	}
+	for (auto &sd : _secrets_settings) {
```
Please put the existing code, except the bit for the company below, in a separate function with `const SettingTable &` as parameter and iterate over the settings in the parameter. Now you're duplicating 20 lines of code just to do exactly the same but with another collection.
Note: I think it is fine to run the full name and short name check on `_settings` first, and then do the same on `_secret_settings` so you only need one function and two calls to it. After all, any setting with a short name that exists in both the _settings as _secret_settings seems like a bug; why is there something with a name that implies it requires secrecy in the non-secret settings?

rubidium42 commented on src/table/secrets_settings.ini:
```diff
@@ -0,0 +1,84 @@
+[post-amble]
+};
+[templates]
+SDTC_BOOL  =  SDTC_BOOL(       $var,        $flags, $guiflags, $def,                              $str, $strhelp, $strval, $proc, $from, $to,        $cat, $extra, $startup),
```
If secrets cannot be passed over the network or saved, maybe replace `$flags` with `$flags | SLF_NOT_IN_SAVE | SLF_NO_NETWORK_SYNC` and `$gui_flags` with `$guiflags | SGF_NETWORK_ONLY` here. Then if someone wants to do it, they need to change the templates which would be super visible in the PR. Furthermore it removes the need to specify it on all the settings separately.

TrueBrain replied:
I have made some drafts for this, but they didn't make it into the PR. Mostly as I first want to know what we want to put in `secrets.cfg`. If we settled on that, we can indeed make this all a bit smarter :D (similar for why there are additional template entries here atm).

rubidium42 commented on src/table/secrets_settings.ini:
```diff
@@ -0,0 +1,84 @@
+[post-amble]
+};
+[templates]
+SDTC_BOOL  =  SDTC_BOOL(       $var,        $flags, $guiflags, $def,                              $str, $strhelp, $strval, $proc, $from, $to,        $cat, $extra, $startup),
```
I wouldn't add templates that are not used yet. That only adds unused code, and if they would ever be needed they can easily be taken from some other location.

rubidium42 reviewed (commented):
Generally I think that the ini not removing settings that are not known to it is not a bad thing. This means switching between development/main/beta/rc and release won't remove all your development/main/beta/rc settings.

However, here I would propose implementing something that does remove them from the settings but only under certain circumstances. For me that would be `IsReleasedVersion()` and `_openttd_newgrf_version > 1 << 28 | 12 << 24`, or any release version 1.12.0 or later. By then it is relatively unlikely that you go back to an earlier version regularly, so it would not burden people switching between development and release builds.

Similarly I think, for the average Joe, it would be better to migrate the settings. Or rather, when it is not set in secrets.ini, take the value from settings.ini (you already have the ini file loaded). The removal of the settings should of course come after this. Both of these are not necessarily part of this PR.

The server-name and client-name are not really secret, it's mostly not useful to have them in a settings file that is used by others, though I'm not really sure that warrants them being secrets. Currently last-joined and servers do not really matter, though if those get join keys that might become slightly more problematic. The server-bind-addresses are not really a problem with respect to secrecy, and other people cannot use that information to fake a server. After all, they do not own that IP, or it is a local IP.
The bans are a bit tricky, though maybe this is the time to consider the separate banlist ini file.
So maybe have a `private_secrets.cfg` and `private_settings.cfg`, where passwords and unique IDs go in the former file and other things that should not be distributed in settings.cfg would go in the latter file such as the (last joined) servers. Yes, private secrets are a bit double in the naming, but I hope that brings over the point that they are really private and secrets so people are less likely to distribute them.

TrueBrain commented:
> Similarly I think, for the average Joe, it would be better to migrate the settings. Or rather, when it is not set in secrets.ini, take the value from settings.ini (you already have the ini file loaded). The removal of the settings should of course come after this. Both of these are not necessarily part of this PR.

I really would like to have the migration in this PR too, but my earlier attempts failed. But I now have an idea how I possibly can pull it off :) Tnx to some rewrites someone did :P

> So maybe have a `private_secrets.cfg` and `private_settings.cfg`, where passwords and unique IDs go in the former file and other things that should not be distributed in settings.cfg would go in the latter file such as the (last joined) servers. Yes, private secrets are a bit double in the naming, but I hope that brings over the point that they are really private and secrets so people are less likely to distribute them.

I like this approach. I will make it so :)

TrueBrain commented:
This current version doesn't work as I want it .. our ini-loader is rather annoying :P

I tried to load first from `openttd.cfg` next from `secrets.cfg`. But .. that doesn't do anything, as every entry in the `_secrets_settings` not found is cleared. So .. yeah .. not done yet :) I have some ideas, but it involves rewriting some more code.

glx22 commented on src/settings.cpp:
```diff
@@ -1668,21 +1670,36 @@ static void GRFSaveConfig(IniFile *ini, const char *grpname, const GRFConfig *li
+
+	proc(openttd_ini, _misc_settings,    "misc",  nullptr, only_startup);
 #if defined(_WIN32) && !defined(DEDICATED)
 	proc(ini, _win32_settings,   "win32", nullptr, only_startup);
```
oups ;)

TrueBrain committed:
Codechange: move network-related settings out of settings.ini

This to prepare the code to split up network-related settings
into private / secrets / generic.

TrueBrain committed:
Change: move sensitive information to secrets.cfg and private information to private.cfg

We often ask people for their openttd.cfg, which now includes their
passwords, usernames, etc. It is easy for people to overlook this,
unwillingly sharing information they shouldn't.

By splitting this information over either private.cfg or secrets.cfg,
we make it more obvious they shouldn't be sharing those files, and
hint to what is inside them.

TrueBrain commented:
Finally had a good idea how to do the migration ... by introducing an `ini-file version` :D I initially tried the newgrf-version, but that would mean people using the nightlies wouldn't get the migration. So in the end, I considered it is best to just have a version to indicate if we should do the migration or not.

That does mean that if you go back to before this PR, those fields reset. But if you go past this PR again, as `ini_version` is not changed, it will not load (or remove) those fields from `openttd.cfg` again. It really is a one-time action.
I think this is the best solution for users, but I am open to other suggestions.

However, I am unsure if adding a IniFileVersion is really needed. I could also just add a flag or indicator. Not sure if this would ever be used again in the future. But I need at least something to know you have been past this PR with your configuration .. so alternatives for this are more than welcome.

Also, there is now `private.cfg` to store private information, like name, servers, etc.

rubidium42 commented on src/settings.cpp:
```diff
@@ -1542,7 +1557,11 @@ static void HandleSettingDescs(IniFile *ini, SettingDescProc *proc, SettingDescP
 
-	proc(ini, _settings,         "patches",  &_settings_newgame, only_startup);
+	for (auto &table : _setting_tables) {
+		/* The name "patch" is a fallback, as every setting should sets its own group. */
```
Shouldn't it be `patches` to be consistent with the next line?

TrueBrain replied:
lol, yes, it should. No clue what I was smoking here, but clearly copy/pasting is too much to ask :D :D :D

rubidium42 commented on src/settings.cpp:
```diff
@@ -1519,9 +1566,12 @@ static void SaveVersionInConfig(IniFile *ini)
 	seprintf(version, lastof(version), "%08X", _openttd_newgrf_version);
 
-	const char * const versions[][2] = {
+	std::string ini_version = std::to_string(INIFILE_VERSION);
```
Not that it is your fault, but sometimes I think preventing "duplicate" code goes a bit too far. In this case someone decided to construct a table and then write a loop to copy the data from that table into the settings, just to prevent two (now three) similar looking lines.

What would the consensus be for something something like the following instead of the current ~15 lines?
```
IniGroup *group = ini->GetGroup("version");
group->GetItem("version_string", true)->SetValue(_openttd_revision);
group->GetItem("version_number", true)->SetValue(fmt::format("{:08X}", _openttd_newgrf_version));
group->GetItem("ini_version", true)->SetValue(std::to_string(INIFILE_VERSION));
```
Yes it is duplicating some code, but preventing a lot more boilerplate for setting up the table and looping over it.

rubidium42 commented on src/settings.cpp:
```diff
@@ -1550,25 +1600,58 @@ static void GRFSaveConfig(IniFile *ini, const char *grpname, const GRFConfig *li
 
-	proc(ini, _currency_settings,"currency", &_custom_currency, only_startup);
-	proc(ini, _company_settings, "company",  &_settings_client.company, only_startup);
+	proc(generic_ini, _currency_settings,"currency", &_custom_currency, only_startup);
```
Existing missing space between _currency_settings, and "currency". Though if it gets touched, maybe fix that as well?

rubidium42 commented on src/settings.cpp:
```diff
@@ -1585,20 +1668,32 @@ static IniFile *IniLoadConfig()
 {
-	IniFile *ini = IniLoadConfig();
+	IniFile *generic_ini = IniLoadConfig();
+	IniFile *private_ini = new IniFile();
```
Can't you just put them on the stack, instead of `new` + `delete`? Would mean some `&` in line 1685 though.

rubidium42 commented on src/settings.cpp:
```diff
@@ -1585,20 +1668,32 @@ static IniFile *IniLoadConfig()
 {
-	IniFile *ini = IniLoadConfig();
+	IniFile *generic_ini = IniLoadConfig();
+	IniFile *private_ini = new IniFile();
```
Most of IniLoadConfig is passing `_list_group_names` to the created IniFile. Looking at those names I see bans, servers and server_bind_addresses which go to the private ini. Shouldn't that constructor be called with the appropriate names for (at least) the private ini? Otherwise that won't get the type of bans set to IGT_LIST and I wouldn't know what kinds of repercussions that would have.
Those `list_group_names` do not seem to do much harm, so maybe a:
```
class ConfigIniFile : public IniFile {
public:
    ConfigIniFile(const std::string &filename) : IniFile(_list_group_names) 
    {
        this->LoadFromDisk(filename, NO_DIRECTORY);
    }
```
and then 
```
ConfigIniFile generic_ini(_config_file);
ConfigIniFile private_ini(_private_file);
ConfigIniFile secrets_ini(_secrets_file);
```
and get rid of the `delete` at the end of the function and of course of IniLoadConfig. Might also mean that you can pass them as reference to HandleSettingDesc, though that might be a bridge too far for this PR.

TrueBrain replied:
I might have taken this a bit to an extreme, but it makes the code a whole lot better and more C++. So yeah, sounds like a good plan :D (well, I write this, but you can't see it yet, till I push the new version .. the suspense!)

rubidium42 commented on src/settings.cpp:
```diff
@@ -1609,27 +1704,59 @@ void LoadFromConfig(bool startup)
 void SaveToConfig()
 {
-	IniFile *ini = IniLoadConfig();
+	IniFile *generic_ini = IniLoadConfig();
```
Similar comments as for `LoadFromConfig`.

TrueBrain committed:
Codechange: C++-ify the usage of IniFile in settings.cpp

Instead of creating the object on heap and use a pointer, create
the object on stack and use a guaranteed-not-null pointer.
The size of IniFile doesn't warrent the forcing to heap.

Additionally, use a subclass instead of a function to do some
initial bookkeeping on an IniFile meant to read a configuration.

TrueBrain committed:
Codechange: simplify SaveVersionInConfig()

Clearly someone really wanted to generalize the function, but
in reality it makes it a lot longer than needed. Let's keep it
simple.

nielsmh commented:
Can I suggest writing a short comment at the top of `secrets.cfg` when it gets created, explaining that these are secrets belonging to the specific installation, and isn't something supposed to be secret from the user. I think some might misunderstand the filename as "these are the secrets of the OpenTTD project and now I can go hack everyone". I don't have any good suggestions for phrasing right now.

TrueBrain commented:
> 
> 
> Can I suggest writing a short comment at the top of `secrets.cfg` when it gets created, explaining that these are secrets belonging to the specific installation, and isn't something supposed to be secret from the user. I think some might misunderstand the filename as "these are the secrets of the OpenTTD project and now I can go hack everyone". I don't have any good suggestions for phrasing right now.

To put IRC here:

I am not too worried that people confuse this `secrets` file with one that would allow access to anything owned by OpenTTD. I give our users a tiny bit more credit ;) But there is validity in your suggestion, as we should make a lot more clear that people should not be sharing `secrets.cfg` with anyone.

Sadly, `IniFile` doesn't really allow that (don't be fouled by the `comment` field .. it is just "the junk between this group and the last", including `\n`s). So what I cooked up is this:

When writing `secrets.cfg` (and `private.cfg`) I add a group first called `secrets` (or `private`), which has a comment before it starts stating what the file is about. This means that as long as the file was empty when we first create it, this group will be on top, and so the comment will be at the top of the file.

The small downside of this approach is, that it will always overwrite this comment, so if people manually alter it, it will be overwritten on next save. This shouldn't be a problem at all, but before this change, ini-files always kept what-ever spacing/comments was between groups.
Edit: I changed my mind, and it now only writes this when the `secrets` group doesn't exist yet. That should prevent changing anything if the user manually makes alterations.

rubidium42 commented on src/settings.cpp:
```diff
@@ -1621,7 +1751,7 @@ StringList GetGRFPresetList()
 	StringList list;
 
-	std::unique_ptr<IniFile> ini(IniLoadConfig());
+	std::unique_ptr<IniFile> ini = std::make_unique<ConfigIniFile>(_config_file);
```
You could also do without the `std::unique_ptr` completely. It would then be like `LoadFromConfig`.
